### PR TITLE
(fix)O3-2197: Unable to remove a patient from a list

### PR DESCRIPTION
**Pull request by jwnasambu**
## Requirements

- [x] This PR has a title that briefly describes the work done including the ticket number. If there is a ticket, make sure your PR title includes a [conventional commit](https://o3-dev.docs.openmrs.org/#/getting_started/contributing?id=your-pr-title-should-indicate-the-type-of-change-it-is) label. See existing PR titles for inspiration.
- [x] My work conforms to the [**OpenMRS 3.0 Styleguide**](https://om.rs/styleguide) and [**design documentation**](https://zeroheight.com/23a080e38/p/880723-introduction).
- [x] I checked for feature overlap with [**existing widgets**](https://om.rs/directory).

## Summary
Created a overflow-menu to enable a user delete a patient from a given patient list

## Screenshots
**This is the output using post method**

Uploading screencast 2023-08-23 4 PM-50-35.mp4…

**This is the output using delete method**

https://github.com/openmrs/openmrs-esm-patient-management/assets/33891016/388ef9a1-7e27-4d88-b52c-cd8f4866786b

## Related Issue
https://issues.openmrs.org/browse/O3-2197

## Other

*None.*

### CODE REVIEW

File: packages/esm-patient-list-app/src/patient-table/overflow-menu.component.tsx [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+  cohortMembershipUuid: string;
+}
+
+const OverflowMenuComponent: React.FC<OverflowMenuComponentProps> = ({ cohortMembershipUuid }) => {
+  const { t } = useTranslation();
+  const layoutType = useLayoutType();
+  const desktopLayout = isDesktop(layoutType);
+
+  // eslint-disable-next-line no-console
+  console.log(cohortMembershipUuid);
```
Comment (ibacher) [inline packages/esm-patient-list-app/src/patient-table/overflow-menu.component.tsx:18]: If you're logging for debugging purposes, it's probably better to put this in the callback than here.

File: packages/esm-patient-list-app/src/api/api-remote.ts [hunk 2]
```diff
@@ -136,6 +136,16 @@ export async function addPatientToList(data: AddPatientData, ac = new AbortContr
   return postData(`${cohortUrl}/cohortmember`, data, ac);
 }
 
+export async function removePatientFromList(cohortMembershipUuid: string, ac = new AbortController()) {
+  return deleteData(
+    `${cohortUrl}/cohortmember/${cohortMembershipUuid}`,
+    {
+      voidReason: '',
```
Comment (ibacher) [inline packages/esm-patient-list-app/src/api/api-remote.ts:143]: We're not voiding the cohortMember, just adding an endDate to the record.

File: packages/esm-patient-list-app/src/patient-table/patient-table.component.tsx [hunk 3]
```diff
@@ -137,6 +142,9 @@ const PatientTable: React.FC<PatientTableProps> = ({
                     {row.cells.map((cell) => (
                       <TableCell key={cell.id}>{cell.value?.content ?? cell.value}</TableCell>
                     ))}
+                    <TableCell className="cds--table-column-menu">
```
Comment (ibacher) [inline packages/esm-patient-list-app/src/patient-table/patient-table.component.tsx:145]:
```suggestion
                    <TableCell>
```

File: packages/esm-patient-list-app/src/api/api-remote.ts (hunk 2 above)
Comment (vasharma05) [reply packages/esm-patient-list-app/src/api/api-remote.ts:144]: Call the `deleteData` instead of `postData` @jwnasambu .

File: packages/esm-patient-list-app/src/patient-table/overflow-menu.component.tsx [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+  }, [cohortMembershipUuid, t]);
+
+  return (
+    <Layer className={styles.layer}>
+      <OverflowMenu
+        data-floating-menu-container
+        ariaLabel="Remove patient from the list"
+        size={desktopLayout ? 'sm' : 'lg'}
+        flipped>
+        <OverflowMenuItem
```
Comment (vasharma05) [inline packages/esm-patient-list-app/src/patient-table/overflow-menu.component.tsx:42]: Add an ExtensionSlot in here, with slot name = "cohort-member-action-items"

**Review by vasharma05 — COMMENTED** [review]
Small corrections. otherwise looks good!

File: packages/esm-patient-list-app/src/api/api-remote.ts (hunk 2 above)
Comment (ibacher) [reply packages/esm-patient-list-app/src/api/api-remote.ts:144]: @vasharma05 Deleting isn't the correct thing in this instance. Removing a patient from a cohort makes them no longer a cohort member, so their membership expires. It doesn't void the membership (that's a different operation).

@jwnasambu POST is better than DELETE; however, we need to post something more like the full existing record with an endDate (and remove the voidReason property)

File: packages/esm-patient-list-app/src/api/api-remote.ts [hunk 5]
```diff
@@ -136,8 +136,19 @@ export async function addPatientToList(data: AddPatientData, ac = new AbortContr
...
+    {
+      endDate: new Date(),
+      voidReason: '',
+    },
+    ac,
+  );
+}
+
+export async function createPatientList(cohort: NewCohortDataPayload, ac = new AbortController()) {
+  return deleteData(
```
Comment (ibacher) [inline packages/esm-patient-list-app/src/api/api-remote.ts:151]:
```suggestion
  return postData(
```

Comment (jwnasambu) [reply packages/esm-patient-list-app/src/api/api-remote.ts:151]: @ibacher and @vasharma05 thanks for the review let share the proposed changes shortly!

File: packages/esm-patient-list-app/src/api/api-remote.ts (hunk 2 above)
Comment (vasharma05) [reply packages/esm-patient-list-app/src/api/api-remote.ts:144]: Hi @ibacher and @dkayiwa !
I tried both the DELETE and POST methods, but POST request didn't make any change.
Do we need to use the DELETE or the POST method?
Thanks!

POST Request:
```
curl --location 'https://dev3.openmrs.org/openmrs/ws/rest/v1/cohortm/cohortmember/90fec1c6-8799-4c65-98ea-9053f1568457' \
--header 'Content-Type: application/json' \
--header 'Cookie: JSESSIONID=375DCF7704E139C04853E6277331448F' \
--data '{
    "endDate": "2023-08-07T12:11:12.558Z"
}'
```

Comment (vasharma05) [reply packages/esm-patient-list-app/src/api/api-remote.ts:144]: P.S. DELETE method is deleting the cohortmember as expected.

File: packages/esm-patient-list-app/src/patient-table/patient-table.component.tsx [hunk 6]
```diff
@@ -67,7 +67,7 @@ const PatientTable: React.FC<PatientTableProps> = ({
     () =>
       patients.map((patient, index) => {
         const row = {
-          id: index,
+          id: patient.uuid,
```
Comment (vasharma05) [inline packages/esm-patient-list-app/src/patient-table/patient-table.component.tsx:70]: This needs to be member uuid, instead of patient uuid.
Thanks!

**Review by vasharma05 — CHANGES_REQUESTED** [review]

File: packages/esm-patient-list-app/src/patient-table/patient-table.component.tsx (hunk 6 above)
Comment (jwnasambu) [reply packages/esm-patient-list-app/src/patient-table/patient-table.component.tsx]: @vasharma05 Thanks for the review. I have fixed the proposed changes.

File: packages/esm-patient-list-app/src/api/api-remote.ts (hunk 2 above)
Comment (vasharma05) [reply packages/esm-patient-list-app/src/api/api-remote.ts]: @dkayiwa , any update here ☝️

**Comment by jwnasambu** [comment]
@ibacher Kindly free to review the proposed changes on this PR please!

File: packages/esm-patient-list-app/src/api/api-remote.ts [hunk 7]
```diff
@@ -105,14 +105,24 @@ export async function getPatientListMembers(cohortUuid: string, ac = new AbortCo
...
+  const currentDate = new Date();
   const searchQuery = results.map((p) => p.patient.uuid).join(',');
   const result = await openmrsFetch(`/ws/fhir2/R4/Patient/_search?_id=${searchQuery}`, {
     method: 'POST',
     signal: ac.signal,
   });
 
   const patients: Array<PatientListMember> = result.data.entry.map((e) => e.resource);
-  return patients;
+  const validPatients = patients.filter((patient) => {
```
Comment (Jexsie) [inline packages/esm-patient-list-app/src/api/api-remote.ts:116]: I think simply doing `  const validPatients = patients.filter((patient) => !patient.endDate);`

Comment (ibacher) [reply packages/esm-patient-list-app/src/api/api-remote.ts:116]: It's not though, because it's possible to add a cohort member with an expiration date, so a cohort member might have an end date but still _currently_ be part of the cohort (there's no UI for this yet; it doesn't mean the feature doesn't exist).

Comment (Jexsie) [reply packages/esm-patient-list-app/src/api/api-remote.ts:116]: gotcha! Thanks @ibacher

File: packages/esm-patient-list-app/src/api/api-remote.ts (hunk 2 above)
Comment (dkayiwa) [reply packages/esm-patient-list-app/src/api/api-remote.ts]: https://issues.openmrs.org/browse/COH-54

File: packages/esm-patient-list-app/src/patient-table/overflow-menu.component.tsx [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+  }, [cohortMembershipUuid, cohortName, mutatePatientListMembers, t]);
+
+  return (
+    <Layer className={overflowMenuStyles.layer}>
+      <OverflowMenu
+        data-floating-menu-container
+        ariaLabel={`Remove patient from the ${cohortName} list`}
+        size={desktopLayout ? 'sm' : 'lg'}
+        flipped>
+        <ExtensionSlot name={'cohort-member-action-items'} />
```
Comment (ibacher) [inline packages/esm-patient-list-app/src/patient-table/overflow-menu.component.tsx:54]:
```suggestion
        <ExtensionSlot name="cohort-member-action-items" />
```

File: packages/esm-patient-list-app/src/patient-table/overflow-menu.component.tsx (hunk 4 above)
Comment (ibacher) [reply packages/esm-patient-list-app/src/patient-table/overflow-menu.component.tsx:55]: @vasharma05 Was the intention here to have the "Remove Patient" stuff as an extension?

File: packages/esm-patient-list-app/src/patient-table/patient-table.component.tsx [hunk 9]
```diff
@@ -61,13 +63,15 @@ const PatientTable: React.FC<PatientTableProps> = ({
...
+  cohortName,
+  mutatePatientListMembers,
 }) => {
   const layout = useLayoutType();
   const rows: Array<any> = useMemo(
     () =>
       patients.map((patient, index) => {
         const row = {
-          id: index,
+          id: patient.membershipUuid,
```
Comment (ibacher) [inline packages/esm-patient-list-app/src/patient-table/patient-table.component.tsx:74]: Why this change?
